File: src/prose/ProseText.cpp
#include "prose/ProseText.h"

#include <algorithm>
#include <cctype>
#include <sstream>

namespace slm
{
    bool ProseText::isProse(const std::string &line)
    {
        if (line.empty())
        {
            return false;
        }
        const char first = line.front();
        return first != '|' && first != '#' && first != '>' &&
               line.rfind("$$", 0) != 0 && !std::isdigit(static_cast<unsigned char>(first));
    }

    int ProseText::wordCount(const std::string &s)
    {
        std::istringstream in(s);
        std::string word;
        int count = 0;
        while (in >> word)
        {
            ++count;
        }
        return count;
    }

    int ProseText::commaCount(const std::string &s)
    {
        return static_cast<int>(std::count(s.begin(), s.end(), ','));
    }

    int ProseText::occurrences(const std::string &haystack, const std::string &needle)
    {
        int count = 0;
        for (std::size_t at = haystack.find(needle); at != std::string::npos;
             at = haystack.find(needle, at + 1))
        {
            ++count;
        }
        return count;
    }

    std::string ProseText::shorten(const std::string &s, std::size_t width)
    {
        std::string flat;
        bool space = false;
        for (char c : s)
        {
            if (c == '\n' || c == ' ')
            {
                space = true;
                continue;
            }
            if (space && !flat.empty())
            {
                flat.push_back(' ');
            }
            space = false;
            flat.push_back(c);
        }
        return flat.size() <= width ? flat : flat.substr(0, width);
    }

    std::vector<std::string> ProseText::splitLines(const std::string &text)
    {
        std::vector<std::string> lines;
        std::istringstream in(text);
        std::string line;
        while (std::getline(in, line))
        {
            lines.push_back(line);
        }
        return lines;
    }

    std::string ProseText::trimmed(const std::string &s)
    {
        const std::size_t from = s.find_first_not_of(" \t\r");
        if (from == std::string::npos)
        {
            return {};
        }
        const std::size_t to = s.find_last_not_of(" \t\r");
        return s.substr(from, to - from + 1);
    }

    std::vector<std::string> ProseText::proseSentences(const std::string &text)
    {
        std::vector<std::string> out;
        std::string paragraph;
        bool insideMath = false;
        const auto flush = [&out, &paragraph]() {
            std::string current;
            for (std::size_t at = 0; at < paragraph.size(); ++at)
            {
                current.push_back(paragraph[at]);
                const bool stop = paragraph[at] == '.' || paragraph[at] == '!' ||
                                  paragraph[at] == '?';
                if (stop && (at + 1 == paragraph.size() || paragraph[at + 1] == ' '))
                {
                    if (current.size() > 12)
                    {
                        out.push_back(current);
                    }
                    current.clear();
                }
            }
            paragraph.clear();
        };
        for (const std::string &line : splitLines(text))
        {
            const std::string flat = trimmed(line);
            if (flat == "$$")
            {
                insideMath = !insideMath;
                continue;
            }
            if (insideMath || !isProse(flat))
            {
                flush();
                continue;
            }
            paragraph += (paragraph.empty() ? "" : " ") + flat;
        }
        flush();
        return out;
    }

    std::string ProseText::finalWord(const std::string &sentence)
    {
        std::istringstream in(sentence);
        std::string word;
        std::string last;
        while (in >> word)
        {
            last = word;
        }
        while (!last.empty() && std::string(".,;:!?*)").find(last.back()) != std::string::npos)
        {
            last.pop_back();
        }
        return last;
    }

}

File: tools/checks/prose_text.py
"""Line- and sentence-level reading of prose, shared by every register rule
that scans it.

A line belongs to the prose only when it is not a table row, a heading, a
blockquote, a display-equation delimiter, or a line opening on a digit. A
sentence is prose accumulated across lines until a stop mark closes it, with
display mathematics folded out rather than scanned.

Mirrors src/prose/ProseText.h/.cpp, with one correction: the stop mark search
now requires the punctuation mark to be followed by whitespace or the end of
the text, the way proseSentences() below already did in the C++ original.
Two of the C++ registers (SentenceLengthRegister, NominalChainRegister) used
a cheaper `rfind('.')` instead and so treated the period inside a cross
reference such as "Bölüm 4.1.1" as a sentence end whenever it shared a line
with the sentence's real closing period, inflating the word count of
whatever came after. That divergence is why this port keeps exactly one
stop-mark implementation and has every caller use it.
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass
class ProseFault:
    rule: str
    line: int
    excerpt: str


_STOP_MARKS = (".", "!", "?")
_EMPHASIS = "*_"


def is_prose(line: str) -> bool:
    if not line:
        return False
    first = line[0]
    return (
        first not in "|#>"
        and not line.startswith("$$")
        and not first.isdigit()
    )


def word_count(s: str) -> int:
    return len(s.split())


def comma_count(s: str) -> int:
    return s.count(",")


def occurrences(haystack: str, needle: str) -> int:
    count = 0
    at = haystack.find(needle)
    while at != -1:
        count += 1
        at = haystack.find(needle, at + 1)
    return count


def shorten(s: str, width: int = 62) -> str:
    flat_parts: list[str] = []
    space = False
    for ch in s:
        if ch in ("\n", " "):
            space = True
            continue
        if space and flat_parts:
            flat_parts.append(" ")
        space = False
        flat_parts.append(ch)
    flat = "".join(flat_parts)
    return flat if len(flat) <= width else flat[:width]


def split_lines(text: str) -> list[str]:
    return text.split("\n")


def trimmed(s: str) -> str:
    return s.strip(" \t\r")


def _closes_sentence(text: str, at: int) -> bool:
    """Whether the stop mark at index `at` actually closes a sentence: it is
    followed, once any closing markdown emphasis (`**bold**`, `_italic_`) is
    skipped over, by whitespace or the end of the text.

    Without the skip, a bold-terminated sentence such as
    "**Katsayı tam olarak bire eşit çıkar.**" never closes, because the
    period is immediately followed by `**` rather than a space, and
    everything after it silently merges into the same "sentence".
    """
    if text[at] not in _STOP_MARKS:
        return False
    j = at + 1
    while j < len(text) and text[j] in _EMPHASIS:
        j += 1
    return j == len(text) or text[j] == " "


def last_sentence_stop(text: str) -> int:
    """Index of the last stop mark that actually closes a sentence, per
    `_closes_sentence`, or -1 if there is none.

    A period inside a decimal or a section number such as "4.1.1" is not
    followed by whitespace and so is skipped, which is exactly the
    distinction the buggy C++ callers were missing.
    """
    for i in range(len(text) - 1, -1, -1):
        if _closes_sentence(text, i):
            return i
    return -1


def prose_sentences(text: str) -> list[str]:
    out: list[str] = []
    paragraph = ""
    inside_math = False

    def flush() -> None:
        nonlocal paragraph
        current = ""
        for at, ch in enumerate(paragraph):
            current += ch
            if _closes_sentence(paragraph, at):
                if len(current) > 12:
                    out.append(current)
                current = ""
        paragraph = ""

    for line in split_lines(text):
        flat = trimmed(line)
        if flat == "$$":
            inside_math = not inside_math
            continue
        if inside_math or not is_prose(flat):
            flush()
            continue
        paragraph = (paragraph + " " + flat) if paragraph else flat
    flush()
    return out


def final_word(sentence: str) -> str:
    words = sentence.split()
    if not words:
        return ""
    last = words[-1]
    while last and last[-1] in ".,;:!?*)":
        last = last[:-1]
    return last
